src/VisualStudio/Core/Def/Implementation/DebuggerIntelliSense/DebuggerIntellisenseFilter.cs

```diff
@@ -17,26 +18,53 @@
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.DebuggerIntelliSense
 {
-    internal class DebuggerIntelliSenseFilter<TPackage, TLanguageService> : AbstractVsTextViewFilter<TPackage, TLanguageService>, IDisposable
+    internal class DebuggerIntelliSenseFilter<TPackage, TLanguageService> : AbstractVsTextViewFilter<TPackage, TLanguageService>, IDisposable, IFeatureController
         where TPackage : AbstractPackage<TPackage, TLanguageService>
         where TLanguageService : AbstractLanguageService<TPackage, TLanguageService>
     {
         private readonly ICommandHandlerServiceFactory _commandFactory;
-        private readonly IWpfTextView _wpfTextView;
+        private readonly IFeatureServiceFactory _featureServiceFactory;
         private AbstractDebuggerIntelliSenseContext _context;
         private IOleCommandTarget _originalNextCommandFilter;
-
-        internal bool Enabled { get; set; }
+        private IFeatureDisableToken _completionDisabledToken;
 
         public DebuggerIntelliSenseFilter(
             AbstractLanguageService<TPackage, TLanguageService> languageService,
             IWpfTextView wpfTextView,
             IVsEditorAdaptersFactoryService adapterFactory,
-            ICommandHandlerServiceFactory commandFactory)
+            ICommandHandlerServiceFactory commandFactory,
+            IFeatureServiceFactory featureServiceFactory)
             : base(languageService, wpfTextView, adapterFactory, commandFactory)
         {
-            _wpfTextView = wpfTextView;
             _commandFactory = commandFactory;
+            _featureServiceFactory = featureServiceFactory;
+        }
+
+        internal void EnableCompletion()
+        {
+            var featureService = _featureServiceFactory.GetOrCreate(WpfTextView);
+            if (!featureService.IsEnabled(PredefinedEditorFeatureNames.Completion))
+            {
+                if (_completionDisabledToken == null)
+                {
+                    return;
+                }
+
+                _completionDisabledToken.Dispose();
+                _completionDisabledToken = null;
+
+                // open the document
+                _context.ContextBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
```

Comment:
This is a pure function....this can definitely be deleted.

Reply:
(although whatever you *think* it's doing it's not...)

Reply:
It seems to work. Thank you!

---
In reply to: [270232945](https://github.com/dotnet/roslyn/pull/32631#discussion_r270232945) [](ancestors = 270232945)